### DISCUSSION TITLE
Possible to train the backbone?

### DISCUSSION THREAD

**Discussion by arjun-kg** [open-edge-platform/anomalib#1138] [Q&A]
I was just wondering if it's possible to train the backbone as well. In my understanding, when we set `pretrained: true`, pretrained imagenet weights are loaded in for eg. Resnet. And even if we set `pretrained: false`, the weights seem to be randomly initialized and frozen, and we get a feature extractor that is not useful unless we separately load other weights offline. Is this accurate? 

Is it possible to train the backbone as well while training the anomaly model, i.e., backpropagate all the way through the backbone network during training?

**Comment by samet-akcay** [comment]
It is possible, but depends on the design of the algorithm. `pretrained` flag only determines whether to use the pre-trained weights of the backbone. 

Some algorithms freeze the backbone and use it as a feature extractor, while others continue to fine-tune the backbone. For example, memory-bank based approaches such as PADIM and Patchcore do not train the backbone. 
https://github.com/openvinotoolkit/anomalib/blob/161d66b7aef316f4e847ef4365ca10388762616b/src/anomalib/models/patchcore/torch_model.py#L65-L66

Some other models contiue to train the backbone. For example, knowledge distillation models such as student-teacher, train the backbone of the student model further, as shown below:
https://github.com/openvinotoolkit/anomalib/blob/161d66b7aef316f4e847ef4365ca10388762616b/src/anomalib/models/stfpm/torch_model.py#L34-L41

Overall, it is based on the design of the algorithm itself. Let us know if you have any other questions. Thanks!

**Reply by arjun-kg** [reply]
Thank you for your detailed reply! This means if I use PADIM or PatchCore with `pretrained=False`, that would just be using a fixed backbone with random weights (which basically creates unusable features)?  

And just checking, if the below code means that weights are frozen for Fastflow too? 

https://github.com/openvinotoolkit/anomalib/blob/0c7444e5552241df0fbda8e41cbc89874b5ec604/src/anomalib/models/fastflow/torch_model.py#L155-L156

Suppose the pretrained Imagenet features are not good for my specific dataset, what would you recommend? It seems it's not possible to train something like Fastflow end-to-end including the feature extractor.

**Comment by samet-akcay** [comment]
I'm moving this to the Q&A section in Discussions. We could continue our discussion there.
